### DISCUSSION TITLE
How Can I Implement An Interactive Prompt

### DISCUSSION THREAD

**Discussion by ghost** [remkop/picocli#1769] [Q&A] [answered]
**How Can I Implement An Interactive Prompt ?**
*picocli* has everything that a modern cli framework must have, and I really appreciate @remkop for his great work, people think that the only good option nowadays for creating a good cli application is *golang* but they underestimate the power of Java + PicoCLI + GraalVM :)

but unfortunately, I was unable to find the library for implementing an interactive prompt for my application something like [PromptUI](https://github.com/manifoldco/promptui) in *golang* if we add the power of interactive prompt to picocli or as a separate library it would be amazing 💡

if there is a library for this purpose I will be very appropriated if someone introduced to me
if there isn't I will be very happy to help, just give me something keyword to search and I will find a way to implemented (it would be hard for sure 💯)

**Comment by adriens** [comment]
I could add also this approach https://github.com/charmbracelet/gum

**Comment by remkop** [comment, marked as answer]
Picocli offers integration with JLine ([jline 2](https://github.com/remkop/picocli/tree/main/picocli-shell-jline2) and [jline 3](https://github.com/remkop/picocli/tree/main/picocli-shell-jline3)) for shell-like interactive interfaces where users type commands.

You may be interested in the [Lanterna project](https://github.com/mabe02/lanterna), which offers a TUI (text UI), sort of like rendering components like checkboxes, buttons, dialogs etc in a text terminal.

The PromptUI projection you pointed to seems to be a hybrid of these.
I don't know what the state of the art is in this area for Java-based projects/libraries.
I did some research two years ago, and listed my findings in [this ticket](https://github.com/remkop/picocli/issues/633); it mentions some other libraries doing work on interactive, hybrid and "fluid" consoles. If you find new ones I would appreciate it if you could add links to that ticket, to help me keep knowledge on this topic in one place. ;-)

For completeness, some more ideas for richer terminal integration are mentioned in this ticket: https://github.com/remkop/picocli/issues/729
